File: basics/jvm_based/src/main/scala/scalaprograms/sorting/MergeSort.scala
package scalaprograms.sorting

/**
  * 1. This sorting technique has a worst time complexity of O(n * log n).
  * 2. Stable sort.
  *
  * * One disadvantage of this sorting technique is that
  * * it creates two array objects for each merge operation.
  * * Line number 38.
  */
object MergeSort extends Sort {

  def main(args: Array[String]): Unit = {
    val unsorted = Seq(1, 23, 4, 100, 6, 8, 94, 2)
    println(s"Original order of elements: $unsorted")
    // ordering defines the sorting to be in descending order.
    val sorted = sort(unsorted)(Ordering.Int.reverse)
    println(s"Sorted order of elements: $sorted")
  }

  /**
    * This implementation create one extra object for each call
    * to [[merge()]] method for returning the result,
    * than the imperative code in Java.
    */
  private def merge[T](xs1: Seq[T], xs2: Seq[T])(ord: Ordering[T]): Seq[T] = {
    (xs1, xs2) match {
      case (l1, Nil) => l1
      case (Nil, l2) => l2
      case (h1 +: t1, h2 +: _) if ord.lt(h1, h2) => h1 +: merge(t1, xs2)(ord)
      case (h1 +: _, h2 +: t2) if ord.gt(h1, h2) => h2 +: merge(xs1, t2)(ord)
    }
  }

  override def sort[T](in: Seq[T])(implicit ord: Ordering[T]): Seq[T] = {
    val n = in.length / 2
    if (n == 0) in
    else {
      val (left, right) = in splitAt n
      merge(sort(left)(ord), sort(right)(ord))(ord)
    }
  }
}

File: basics/jvm_based/src/main/java/javaprograms/sorting/MergeSort.java
package javaprograms.sorting;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;

/**
 * 1. This sorting technique has a average time complexity of O(n * log n).
 * 2. Stable sort.
 *
 * One disadvantage of this sorting technique is that
 * it creates two array objects for each merge operation.
 * Line number 28 ad 29.
 */
public class MergeSort extends Sort {

    public static void main(String[] args) {
        ArrayList<Integer> arr = new ArrayList<>(Arrays.asList(1, 23, 4, 100, 6, 8, 94, 2));
        System.out.println("Original order of elements: " + arr.toString());
        // Comparator for the sorting to be in descending order.
        new MergeSort().sort(arr, Comparator.comparingInt(o -> -o));
        System.out.println("Sorted order of elements: " + arr.toString());
    }

    private <T> void merge(ArrayList<T> arr, int left, int mid, int right, Comparator<T> comparator) {
        int n1 = mid - left + 1;
        int n2 = right - mid;
        // creating two temporary arrays
        ArrayList<T> l_arr = new ArrayList<>();
        ArrayList<T> r_arr = new ArrayList<>();
        for (int i = 0; i < n1; ++i)
            l_arr.add(i, arr.get(left + i));
        for (int j = 0; j < n2; ++j)
            r_arr.add(j, arr.get(mid + 1 + j));
        // merge the two sub arrays in order.
        int i = 0, j = 0;
        int k = left;
        while (i < n1 && j < n2) {
            if (comparator.compare(l_arr.get(i), r_arr.get(j)) < 0) {
                arr.set(k, l_arr.get(i));
                i++;
            } else {
                arr.set(k, r_arr.get(j));
                j++;
            }
            k++;
        }
        // copy the left over
        while (i < n1) {
            arr.set(k, l_arr.get(i));
            i++;
            k++;
        }
        while (j < n2) {
            arr.set(k, r_arr.get(j));
            j++;
            k++;
        }
    }

    private <T> void divide(ArrayList<T> arr, int left, int right, Comparator<T> comparator) {
        if (left < right) {
            int mid = (left + right) / 2;
            divide(arr, left, mid, comparator);
            divide(arr, mid + 1, right, comparator);
            merge(arr, left, mid, right, comparator);
        }
    }

    public <T> void sort(ArrayList<T> in, Comparator<T> comparator) {
        int length = in.size();
        divide(in, 0, length - 1, comparator);
    }
}
